portfolio/src/main/java/com/google/sps/servlets/ImageUploadServlet.java

```diff
@@ -0,0 +1,46 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+
+/**
+* Servlet that handles requests to upload files using Blobstore
+*/
+@WebServlet("/comment-image-uploads")
+public class ImageUploadServlet extends HttpServlet{
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
You should probably switch the form action to `POST`, and this to `doPost`.

more detail: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods

The practical difference is that browsers will never automatically retry a failed POST request, and will warn you before re-submitting a POST request if you refresh the page (if you've ever seen a "Confirm Form Resubmission" dialog, that's why).

`GET` is generally only used for forms that *don't* change the server state, e.g. search.

Reply:
I probably should have renamed this servlet and specified in the javadocs that this servlet returns a url that allows users to upload a file to Blobstore. The fetchBlobstoreURL() function in script.js fetches this url and then sets that as the action for the comments form (this function is called when index.html loads on the screen). The doPost method in DataServlet.java gets the url of the image the user uploaded and saves that with the other properties for a comment in datastore. Also the method for this form is already set to POST.

Should I keep this method, or should I still change this to doPost?

Reply:
Also, I'll go ahead and update the javadocs and rename this servlet.

Reply:
Ahh makes sense, thanks!